agent/lib_composer.c

```diff
@@ -0,0 +1,233 @@
+    nrl_verbosedebug(NRL_INSTRUMENT,
+                     "%s: zend_eval_string(%s) failed, result=%d", __func__,
+                     code, result);
+    nr_free(code);
+    return NR_FAILURE;
+  }
+
+  zval_dtor(&retval);
+  nr_free(code);
+
+  // Make sure runtime API is available after loading
+  // Composer\\InstalledVersions class:
+  if (!nr_execute_handle_autoload_composer_is_initialized()) {
+    nrl_verbosedebug(NRL_INSTRUMENT,
+                     "%s: unable to initialize Composer runtime API", __func__);
+    return NR_FAILURE;
+  }
+
+  return NR_SUCCESS;
+}
+
+static void nr_execute_handle_autoload_composer_get_packages_information(
+    const char* vendor_path) {
+  zval retval;  // This is used as a return value for zend_eval_string.
+                // It will only be set if the result of the eval is SUCCESS.
+  int result = -1;
+
+  // nurunlikely because this should alredy be ensured by the caller
+  if (nrunlikely(!NRINI(vulnerability_management_package_detection_enabled))) {
+    // do nothing when collecting package information for vulnerability
+    // management is disabled
+    return;
+  }
+
+  // nurunlikely because this should alredy be ensured by the caller
+  if (nrunlikely(!NRINI(vulnerability_management_composer_api_enabled))) {
+    // do nothing when use of composer to collect package info is disabled
+    return;
+  }
+
+  // clang-format off
+  char* getallrawdata
+        = ""
+        "(function() {"
+        "  try {"
+        "    $root_package = \\Composer\\InstalledVersions::getRootPackage();"
+        "    $packages = array();"
+        "    foreach (\\Composer\\InstalledVersions::getAllRawData() as $installed) { "
+        "      foreach ($installed['versions'] as $packageName => $packageData) {"
+        "        if (is_array($root_package) && array_key_exists('name', $root_package) && $packageName == $root_package['name']) {"
+        "          continue;"
+        "        }"
+        "        if (isset($packageData['pretty_version'])) {"
+        "          $packages[$packageName] = ltrim($packageData['pretty_version'], 'v');"
+        "        }"
+        "      }"
+        "    }"
+        "    return $packages;"
+        "  } catch (Exception $e) {"
+        "    return NULL;"
+        "  }"
+        "})();";
+  // clang-format on
+
+  if (NR_SUCCESS != nr_execute_handle_autoload_composer_init(vendor_path)) {
+    nrl_debug(NRL_INSTRUMENT,
+              "%s - unable to initialize Composer runtime API - package info "
+              "unavailable",
+              __func__);
+    return;
+  }
+
+  nrl_verbosedebug(NRL_INSTRUMENT, "%s - Composer runtime API available",
+                   __func__);
+
+  result
+      = zend_eval_string(getallrawdata, &retval, "composer_getallrawdata.php");
+  if (SUCCESS != result) {
+    nrl_verbosedebug(NRL_INSTRUMENT, "%s - composer_getallrawdata.php failed",
+                     __func__);
+    return;
+  }
+  if (IS_ARRAY == Z_TYPE(retval)) {
```

Comment:
could also consider checking for IS_NULL so we know if an exception occurred.

Reply:
@zsistla Will printing the returned value with `nr_format_zval_for_debug` in `else` clause [here](https://github.com/newrelic/newrelic-php-agent/blob/feat/vm/user-composer-for-package-info/agent/lib_composer.c#L151-L154) be good enough?

Reply:
Good idea, that would be great.

Reply:
Implemented in af1170a5c377631bbb238bb48b527a7310935503.